### PR TITLE
Refactor Ernie 4.5's MoE

### PR DESCRIPTION
**Pull request by LysandreJik**

### CODE REVIEW

File: src/transformers/models/deepseek_v2/modular_deepseek_v2.py
Comment (LysandreJik) [inline src/transformers/models/deepseek_v2/modular_deepseek_v2.py:1]: Ignore this file

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 1]
```diff
@@ -96,7 +96,46 @@ def forward(self, hidden_states):
...
-# FIXME: update the expert
+class Ernie4_5_MoeNaiveMoe(nn.ModuleList):
+    def __init__(self, config):
+        super().__init__()
+        self.top_k = config.num_experts_per_tok
+        self.num_experts = config.num_local_experts
+        for _ in range(self.num_experts):
+            self += [Ernie4_5_MoeMLP(config, config.moe_intermediate_size)]
+
+    def forward(self, hidden_states, e_score_correction_bias, device_type):
```
Comment (LysandreJik) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: This MoE is different as it has an `e_score_correction_bias` which gets added after the Softmax, in order to select the expert

Rather than splitting the Naive MoE into two, I chose to pass the `e_score_correction_bias` directly to the method.

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 2]
```diff
@@ -152,37 +189,7 @@ def forward(
...
-            # Index the correct hidden states and compute the expert hidden state for
-            # the current expert. We need to make sure to multiply the output hidden
-            # states by `routing_weights` on the corresponding tokens (top-1 and top-2)
-            current_state = hidden_states[None, top_x].reshape(-1, hidden_dim)
-            current_hidden_states = expert_layer(current_state) * routing_weights[top_x, idx, None]
-
-            # However `index_add_` only support torch tensors for indexing so we'll use
-            # the `top_x` tensor here.
-            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+        final_hidden_states = self.experts(hidden_states, self.moe_statics.e_score_correction_bias.squeeze(), device_type)
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: Check with TP, might fail - was one of the reasons I moved it to its own forward (might no longer be necessary)

Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: Meaning `self.moe_statics.e_score_correction_bias.squeeze()`

Comment (LysandreJik) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: you have eyes everywhere :) just pinged Arthur on it because that comment made me uneasy

Comment (ArthurZucker) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: yep if you have more details @vasqu I think it should be fine !

Comment (ArthurZucker) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: this will break your EP implementation but it was already going agains our standards to not have an explicit EP path!

Comment (ArthurZucker) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: the only issue I see here is for @hmellor as this will ad an extra arg but he probably protected them already!

Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: Iirc, there was a hook attached to the parameter (ie your module weights were correctly converted but the parameter stayed as dtensor) and when you attempted anything on it, it would break - can check later if it indeed does break

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 3]
```diff
@@ -343,38 +380,10 @@ def forward(
             # router_logits: (batch * sequence_length, n_experts)
             router_logits = self.gate(hidden_states.float())
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:383]: Gate isnt used anymore and the correction bias needs to be under the top k for a bias

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 4]
```diff
@@ -288,6 +287,46 @@ def forward(self, hidden_states):
...
+        super().__init__()
+        self.top_k = config.num_experts_per_tok
+        self.num_experts = config.num_local_experts
+        for _ in range(self.num_experts):
+            self += [Ernie4_5_MoeMLP(config, config.moe_intermediate_size)]
+
+    def forward(self, hidden_states, e_score_correction_bias, device_type):
+        # router_logits: (batch * sequence_length, n_experts)
+        with torch.autocast(device_type=device_type, enabled=False):  # Force float32
+            routing_weights = F.softmax(hidden_states, dim=1, dtype=torch.float)
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py]:
```suggestion
            routing_weights = F.softmax(routing_logits, dim=1, dtype=torch.float)
```
We need the logits from the gate here, if we move things

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 5]
```diff
@@ -288,6 +287,46 @@ def forward(self, hidden_states):
...
+            _, selected_experts = torch.topk(routing_weights + e_score_correction_bias, self.top_k, dim=-1)
+            routing_weights = torch.gather(routing_weights, dim=-1, index=selected_experts)
+            routing_weights = routing_weights / torch.clamp(
+                routing_weights.sum(dim=-1, keepdim=True), min=self.norm_min
+            )
+            routing_weights = routing_weights.to(hidden_states.dtype)
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:307]: I.e. the gate should be another module that takes the bias and returns the topk, etc

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: TP wasnt an issue anymore! Resolving this

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:307]: for vllm that's not helpful!

**Comment by LysandreJik** [comment]
@vasqu feel free to takeover and push directly here in case you want to do deeper fixes; need to fix one or two other models before the release. Thanks!

**Comment by vasqu** [comment]
Got it on it @LysandreJik

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py [hunk 6]
```diff
@@ -152,43 +191,18 @@ def forward(
...
-            )
-            routing_weights = routing_weights.to(hidden_states.dtype)
-
-        final_hidden_states = torch.zeros(
-            (batch_size * sequence_length, hidden_dim), dtype=hidden_states.dtype, device=hidden_states.device
+        final_hidden_states = self.experts(
+            hidden_states,
+            routing_weights=router_logits,
+            routing_bias=self.moe_statics.e_score_correction_bias.squeeze(),
+            device_type=device_type,
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py:198]: Needed to pass them as kwarg for whatever reason :sweat_smile:

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 7]
```diff
@@ -67,7 +66,7 @@ def __init__(self, config, intermediate_size=None):
         super().__init__()
         self.config = config
         self.hidden_size = config.hidden_size
-        self.intermediate_size = intermediate_size if intermediate_size is not None else config.intermediate_size
+        self.intermediate_size = config.intermediate_size if intermediate_size is None else intermediate_size
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:69]: Modular copy

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 8]
```diff
@@ -457,11 +447,9 @@ def forward(
         residual = hidden_states
         hidden_states = self.post_attention_layernorm(hidden_states)
         hidden_states = self.mlp(hidden_states)
-        # For the MoE layers, we need to unpack
         if isinstance(hidden_states, tuple):
-            hidden_states, _ = hidden_states
+            hidden_states = hidden_states[0]
         hidden_states = residual + hidden_states
-
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:464]: Also modular

File: src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py [hunk 9]
```diff
@@ -288,6 +287,48 @@ def forward(self, hidden_states):
         return hidden_states + self.e_score_correction_bias.squeeze()
 
 
+class Ernie4_5_MoeNaiveMoe(nn.ModuleList):
+    def __init__(self, config):
+        super().__init__()
+        self.top_k = config.moe_k
+        self.num_experts = config.moe_num_experts
```
Comment (vasqu) [inline src/transformers/models/ernie4_5_moe/modeling_ernie4_5_moe.py:294]: I had to rename here, is there a specific renaming going on?

File: src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py (hunk 2 above)
Comment (hmellor) [reply src/transformers/models/ernie4_5_moe/modular_ernie4_5_moe.py]: I think I can add a forward hook to my `FusedMoE` in vLLM to ignore extra arguments. The `e_score_correction_bias` is passed to `FusedMoE` on init so it'll still be handled.
